module/VuFind/src/VuFind/RateLimiter/Turnstile/Turnstile.php

```diff
@@ -0,0 +1,164 @@
+            'https://challenges.cloudflare.com/turnstile/v0/siteverify';
+        $body = [
+            'secret' => $secretKey,
+            'response' => $token,
+        ];
+        $response = $this->httpService->post(
+            $url,
+            json_encode($body),
+            'application/json'
+        );
+
+        if ($response->isOk()) {
+            $responseData = json_decode($response->getBody(), true);
+            $success = $responseData['success'];
+        } else {
+            // Unexpected error.  Treat as a positive result, since it's not the user's fault.
+            $this->logWarning('Verification process failed, allowing traffic: '
+                . $response->getStatusCode() . $response->getBody());
+            $success = true;
+        }
+
+        return $success;
+    }
+
+    /**
+     * Validate a token and save the result
+     *
+     * @param string $token    The token generated by the Turnstile widget
+     * @param string $policyId RateLimiterManager policy ID
+     * @param string $clientIp Client IP address
+     *
+     * @return bool Whether validation was successful
+     */
+    public function validateAndCacheResult($token, $policyId, $clientIp)
+    {
+        $success = $this->validateToken($token);
+        $this->setTurnstileResult($policyId, $clientIp, $success);
+        return $success;
+    }
+
+    /**
+     * Check for a prior, cached result from Turnstile under this client IP and policy.
+     *
+     * @param string $policyId The policy ID
+     * @param string $clientIp The client IP
+     *
+     * @return ?bool Null if there is no prior result, or if Turnstile is disabled;
+     *               otherwise a boolean representing the Turnstile result.
+     */
+    public function checkPriorTurnstileResult($policyId, $clientIp)
+    {
+        if (!($this->config['Policies'][$policyId]['turnstileRateLimiterSettings'] ?? false)) {
+            return null;
+        }
+        $cacheKey = $this->getTurnstileCacheKey($policyId, $clientIp);
+        return $this->turnstileCache->getItem($cacheKey);
+    }
+
+    /**
+     * Store a Turnstile result for this client IP and policy.
+     *
+     * @param string $policyId The policy ID
+     * @param string $clientIp The client IP
+     * @param bool   $success  The result to store
+     *
+     * @return void
+     */
+    protected function setTurnstileResult($policyId, $clientIp, $success)
+    {
+        $cacheKey = $this->getTurnstileCacheKey($policyId, $clientIp);
+        $this->turnstileCache->setItem($cacheKey, $success);
+    }
+
+    /**
+     * Generate a key for the Turnstile cache.
+     *
+     * @param string $policyId The policy ID
+     * @param string $clientIp The client IP
+     *
+     * @return string The cache key
+     */
+    protected function getTurnstileCacheKey($policyId, $clientIp)
+    {
+        $key = $policyId . '--' . $clientIp;
+        $key = str_replace('.', '-', $key);
```

Comment:
Demian suggested using a hash here...in progress.

Reply:
Actually I think I like having unhashed keys here -- readable keys makes the cache itself easier to inspect/debug if need be.

Reply:
No objections here; that was a suggestion and not a demand. :-)